### PR TITLE
SkinnedMesh: Apply skinning while raycasting

### PR DESCRIPTION
**Pull request by donmccurdy**
This PR just fixes the remaining feedback on https://github.com/mrdoob/three.js/pull/8953. Compared to the original PR from 2016:

- Support for all BufferAttribute and InterleavedBufferAtribute options
- No Geometry support (SkinnedMesh no longer supports Geometry)
- Zero object allocation during raycasting

As you might imagine, doing skinning while raycasting is slower than not doing skinning while raycasting. I think that is acceptable: there isn't much use for the latter. The object's bounding box is _not_ affected by skinning, but I believe that can be left as a future improvement. Being able to tell users "you need to scale your bounding box up a bit" is a much better answer than not supporting raycasting at all. 🙂

### CODE REVIEW

**Comment by Mugen87** [comment]
> Being able to tell users "you need to scale your bounding box up a bit" is a much better answer than not supporting raycasting at all. 

Agreed. Especially since enhancing the AABB is definitely easier for users than implementing the raycasting code. So as a first step, I think the PR is acceptable. Although the bounding volume issue still needs to be solved for a more reliable solution.

**Review by Mugen87 — APPROVED** [review]

**Comment by mrdoob** [comment]
Looks good!

File: src/objects/Mesh.js [hunk 1]
```diff
@@ -426,6 +426,14 @@ function checkBufferGeometryIntersection( object, material, raycaster, ray, posi
 
 	}
 
+	if ( object.boneTransform ) {
```
Comment (mrdoob) [inline src/objects/Mesh.js:429]: How about this?

```js
if ( object.isSkinnedMesh ) {
```

Comment (donmccurdy) [reply src/objects/Mesh.js]: Done ✅

**Comment by mrdoob** [comment]
Thanks!

**Comment by zlf2002171** [comment]
unsubscribe

2020-05-18 

zlf2002171 



发件人：Michael Herzog <notifications@github.com>
发送时间：2020-04-25 05:15
主题：Re: [mrdoob/three.js] SkinnedMesh: Apply skinning while raycasting (#19178)
收件人："mrdoob/three.js"<three.js@noreply.github.com>
抄送："Subscribed"<subscribed@noreply.github.com>

@Mugen87 approved this pull request.

**Comment by makc** [comment]
see, [in 8953 the input vector and the result vector were two different things](https://github.com/mrdoob/three.js/pull/8953/commits/9ac4df6fdd17c5e269e5a44247e3a56d28449fb4#diff-6ce90d9453a7842ca9e0630fedd6dfeeecbb49d253403bca6497d277f39fe8f4R201), but here they are conflated. So, the fix would be changing
```
_basePosition.fromBufferAttribute( geometry.attributes.position, index ).applyMatrix4( this.bindMatrix );
```

to
```_basePosition.copy(target).applyMatrix4( this.bindMatrix );
```

@donmccurdy ?

**Comment by donmccurdy** [comment]
Your change looks right to me – if you have a way to test this change with morph targets and skinning a PR would be great. 👍

**Comment by makc** [comment]
@donmccurdy here is a test case https://jsfiddle.net/2z04ya9s/

**Comment by kzka90** [comment]
I arrived here because I'm trying to RayCast a model loaded from the GLTF loader.

Reading thru all these, it seems that it is not being detected by the RayCaster because it loads as a Skinned Mesh. And I'm a bit lost if there's yet a solution for this.

For my project, I cannot add boundaries manually because GLTFs are loaded from the user's computer and code needs to work agnostically. I cannot ask the users to manually do their boundaries.

**Pull request merged**